### PR TITLE
Add --host-cluster-name to kubefed2 join to override use of kubeconfig context name for resources

### PR DESCRIPTION
**Pull request by pmorie**
Addresses #577

### CODE REVIEW

**Comment by marun** [comment]
This LGTM

@xunpan Can you please take a look?

File: pkg/kubefed2/unjoin.go [hunk 1]
```diff
@@ -110,6 +114,10 @@ func (j *unjoinFederation) Complete(args []string) error {
 		j.clusterContext = j.ClusterName
 	}
 
+	if j.hostClusterName == "" && strings.ContainsAny(j.HostClusterContext, ":/") {
+		glog.Fatal("host-cluster-name must be set if the name of the host cluster context contains one of \":\" or \"/\"")
```
Comment (xunpan) [inline pkg/kubefed2/unjoin.go]: It is preferred to return an error instead of directly `glog.Fatal()`

Comment (pmorie) [reply pkg/kubefed2/unjoin.go]: Valid comments, will incorporate.

**Comment by pmorie** [comment]
@xunpan comments addressed, PTAL

File: pkg/kubefed2/join.go [hunk 2]
```diff
@@ -145,6 +149,10 @@ func (j *joinFederation) Complete(args []string) error {
 		j.clusterContext = j.ClusterName
 	}
 
+	if j.hostClusterName == "" && strings.ContainsAny(j.HostClusterContext, ":/") {
+		glog.Fatal("host-cluster-name must be set if the name of the host cluster context contains one of \":\" or \"/\"")
+	}
+
```
Comment (xunpan) [inline pkg/kubefed2/join.go]: `join` should have similar logic as `unjoin` for messages and `hostClusterName` checking logic.
```
	if j.hostClusterName == "" && strings.ContainsAny(j.HostClusterContext, ":/") {
		return goerrors.New("host-cluster-name must be set if the name of the host cluster context contains one of \":\" or \"/\"")
	}

	if j.hostClusterName != "" && strings.ContainsAny(j.hostClusterName, ":/") {
		return goerrors.New("host-cluster-name may not contain \"/\" or \":\"")
	}
```

**Comment by pmorie** [comment]
@xunpan addressed, PTAL

**Pull request merged**